libs/user/user.ts

```diff
@@ -1,75 +1,78 @@
 import Queue from "queue";
 import path from "path";
-import {
-  CONTRACT_SHIELD,
-  NIGHTFALL_DEFAULT_CONFIG,
-  TX_FEE_DEFAULT,
-} from "./constants";
-import { UserConfig, UserDeposit } from "./types";
+import { CONTRACT_SHIELD, TX_FEE_DEFAULT } from "./constants";
+import { UserFactoryOptions, UserOptions, UserDeposit } from "./types";
 import { Client } from "../client";
 import { Web3Websocket, getEthAddressFromPrivateKey } from "../ethereum";
 import { createZkpKeysFromMnemonic } from "../nightfall";
 import { createDeposit } from "../transactions/deposit";
 import { parentLogger } from "../utils";
+import { createOptions } from "./validations";
 
 const logger = parentLogger.child({
   name: path.relative(process.cwd(), __filename),
 });
 
-class User {
-  // Set by constructor
-  web3Websocket;
-  client;
-
-  // Set by init
-  shieldContractAddress: null | string = null;
-  ethPrivateKey: null | string = null;
-  ethAddress: null | string = null;
-  nightfallMnemonic: null | string = null;
-  zkpKeys: any = null;
-
-  constructor(env = NIGHTFALL_DEFAULT_CONFIG) {
-    logger.debug({ env }, "new User connected to");
-
-    const blockchainWsUrl = env.blockchainWsUrl.toLowerCase();
-    this.web3Websocket = new Web3Websocket(blockchainWsUrl);
+class UserFactory {
+  static async create(options: UserFactoryOptions) {
+    logger.debug("UserFactory :: create");
+    createOptions.validate(options);
 
-    const clientApiUrl = env.clientApiUrl.toLowerCase();
-    this.client = new Client(clientApiUrl);
-  }
-
-  // TODO improve return type
-  async init(config: UserConfig) {
-    logger.debug({ config }, "User :: init");
+    // Instantiate Client and Web3Websocket
+    const client = new Client(options.clientApiUrl.toLowerCase());
+    const web3Websocket = new Web3Websocket(
+      options.blockchainWsUrl.toLowerCase(),
+    );
 
-    // Set this.shieldContractAddress
-    logger.debug("User :: setShieldContractAddress");
-    this.shieldContractAddress = await this.client.getContractAddress(
+    // Get Shield contract address
+    const shieldContractAddress = await client.getContractAddress(
       CONTRACT_SHIELD,
     );
+    if (!shieldContractAddress) return null;
```

Comment:
I think we could raise an exception here since the `shieldContractAddress` is required.